### DISCUSSION TITLE
fusion_indicator

### DISCUSSION THREAD

**Discussion by HPL001** [idaholab/moose#29615] [Q&A Modules: Phase field]
Dear moose developers, 
Hello, I defined a new region with the following expression. According to my logic, the region I defined should be marked as 1, but I don't know why they are all marked as 0.5.
```
[AuxKernels]
  [fusion_indicator]
    type = ParsedAux
    variable = fusion_region
    coupled_variables = 'gr0 gr1 bnds'
    expression = 'if(abs(gr0 - 0.5) < 0.1, if(abs(gr1 - 0.5) < 0.1, 1, 0), 0)'
    execute_on = 'TIMESTEP_END'
  []
[]
```
this is picture of result.
![image](https://github.com/user-attachments/assets/b7efc4e4-7bb7-4728-8b88-c97962b517f3)

**Comment by GiudGiud** [comment]
What variable type is fusion region? Some variables have visualization artefacts in paraview.
Is this paraview or peacock or ?
And did you set the color map scale to go from 0 to 1?

**Reply by HPL001** [reply]
Thank you for your reply. For the above problem, I actually wanted to capture some areas belonging to two particles in this way, and use them as the sintered neck (or GB) area. However, when I encountered this problem, I enlarged the scale to 0-1, which is still the case. I don't know if this is correct, or if I have a better way to deal with the sintered neck.

**Reply by GiudGiud** [reply]
> What variable type is fusion region? Some variables have visualization artefacts in paraview.
Is this paraview or peacock or ?

What about this?

> actually wanted to capture some areas belonging to two particles in this way, and use them as the sintered neck (or GB) area

Did you use an auxiliary variable to try to find these areas?
EDIT: Ok I see below

> I don't know if this is correct, or if I have a better way to deal with the sintered neck.

This looks fine to me. It's just the variable value that's weird

**Comment by HPL001** [comment]
I refer to the determination of the free surface pointed out in the paper of 18 years, and adopt the following code to set up the free surface. Normally, the value of these selected areas should be marked as 1, but here it is 0.5, and I am not sure why. Will it make any difference to my results?
![image](https://github.com/user-attachments/assets/f296c0e4-d2b3-4115-9a7e-4d148371a2ec)
```
[AuxVariables]
  [free_surface_region]
    order = CONSTANT
    family = MONOMIAL
  []
[]
[AuxKernels]
  [free_surface_indicator]
    type = ParsedAux
    variable = free_surface_region
    coupled_variables = 'gr0 gr1 c'
    # expression = 'if(c > 0.25, if(c < 0.75, 1, 0), 0)'
    expression = 'if(c > 0.45, if(c < 0.55, 1, 0), 0)'
    execute_on = 'initial timestep_end'
  []
[]
[Postprocessors]
  [free_surface_area]
    type = ElementIntegralVariablePostprocessor
    variable = free_surface_region
    execute_on = 'initial timestep_end'
  []
[]
```

**Reply by GiudGiud** [reply]
> I refer to the determination of the free surface pointed out in the paper of 18 years

which paper? maybe the authors will have some advice

> Will it make any difference to my results?

It depends. Are you using fusion_region just as an indicator (non-zero indicates the neck)? Or are you relying on its actual value?

```
 [AuxVariables]
  [free_surface_region]
    order = CONSTANT
    family = MONOMIAL
  []
[]
```

ok this makes sense now!
Constant monomials are computed from the integral of the auxkernel evaluation/
So what is happening is that the auxkernel is evaluating to 1 over half the quadrature points and to 0 over the entire half. So the integral amounts to 0.5

**Reply by HPL001** [reply]
I just want to find the area I need to focus on in this way, and then the area of in the post-processing of the calculated (by ElementIntegralVariablePostprocessor). This process has nothing to do with the value of the region, just marking it. However, in moose, I'm not sure that this way of defining variables or post-processing works as well as I thought. Could u give me some advice? thanks.

**Reply by GiudGiud** [reply]
> and then the area of in the post-processing of the calculated (by ElementIntegralVariablePostprocessor).

The problem is that the elements are partially in the fusion region and partially outside, this is why it's 0.5
If you want to compute the area, integrating the variable still works. 0.5 over the element is a better approximation than 1 here

**Reply by HPL001** [reply]
Thank you for your reply. But I think I ran into a numerical problem.Specifically, I've defined a region through this code.
```
[AuxVariables]
  [fusion_region]
    order = CONSTANT
    family = MONOMIAL
  []
[]

[AuxKernels]
  [fusion_indicatorAuxKernel]
    type = ParsedAux
    variable = fusion_region
    coupled_variables = 'gr0 gr1 bnds c'
    expression = 'if(gr0 > 0.45, if(gr0 < 0.55, if(gr1 > 0.45, if(gr1 < 0.55, 1, 0), 0), 0), 0)'
    execute_on = 'initial timestep_end'
  []
[]

[Postprocessors]
  [fusion_area]
    type = ElementIntegralVariablePostprocessor
    variable = fusion_region
    execute_on = 'initial timestep_end'
  []
[]
```
Below I'll show you three consecutive time steps at run time, along with the corresponding area line chart. 
![image](https://github.com/user-attachments/assets/8f65c58e-0852-410e-bf08-a75e2c5f98cc)
![image](https://github.com/user-attachments/assets/6c41c27e-fb92-466d-948c-c67b21a2dc23)
![image](https://github.com/user-attachments/assets/deafdb25-ca2a-4019-954c-3ab5edddefc2)

From the three consecutive graphs, we can find that the area of the selected area is almost growing by 2 small grids at a time. But the area of the line output in peacock increases dramatically in the third picture (Area in red box). This seems to be due to the increase in the number of markers in the third graph. (In fact, I'm not sure if that's the case, so I need your help.)
![image](https://github.com/user-attachments/assets/d5803992-945c-402c-9786-4bec3de3933b)

How should I set up to solve this problem? (I just want to find the areas that meet the filter criteria and then count the area of the area) Looking forward to your reply.

**Reply by GiudGiud** [reply]
Can you try to regenerate this plot with 2x more time steps (dt -> dt/2) ? and see if the results we are seeing are converged in time

**Reply by HPL001** [reply]
Sorry for my late reply. This is the result of redrawing. (The time of the above screenshot is t=76, 83, 91)
![image](https://github.com/user-attachments/assets/4d41c4a7-c785-4fd4-8d3f-4815056a5ab9)

**Reply by GiudGiud** [reply]
Did you make the time step twice smaller? This looks the same to me

**Reply by HPL001** [reply]
Do you mean reduced drawing time or simulation time? I used time step adaptive processing. Also, like this, is there a way for me to force a value to the flagged position? (For example, the value assigned here is 1 instead of 0.5)

**Reply by GiudGiud** [reply]
Use a fixed dT or at least a functionDT

If you don't make the time steps smaller you cannot know whether your results are converged.

**Reply by HPL001** [reply]
Thank you for your reply. I have adopted a fixed time step of 0.01 for calculation. The following is the result diagram.
![image](https://github.com/user-attachments/assets/39e3992a-612c-43b9-aed8-d8810e258b32)
I have a doubt. You mentioned in your last reply that it is difficult to judge whether the time step is convergent because it is not fixed. I don't understand that the adaptive time step is automatically adjusted according to the convergence. Should I turn on adaptive step size in normal calculations?
And, I want to know is there a way for me to force a value to the flagged position? (For example, the value assigned here is 1 instead of 0.5)

**Reply by GiudGiud** [reply]
> . I don't understand that the adaptive time step is automatically adjusted according to the convergence.

Not the same convergence. It's set based on the convergence of the nonlinear solver. That's very different from the time discretization scheme convergence. 


> Should I turn on adaptive step size in normal calculations?

only if you know that the time discretization error is not too large

> And, I want to know is there a way for me to force a value to the flagged position? (For example, the value assigned here is 1 instead of 0.5)

of the fusion variable? That does not seem right to be forcing it

**Reply by HPL001** [reply]
Thank you for your reply. I think there should be a sudden divergence in my first calculation. Since I used the adaptive time step, maybe I crossed this divergence point in the calculation, which caused the data anomaly.

**Comment by HPL001** [comment]
Hello, I now define variables in the following way. I have a question about how to calculate in the postprocessor(ElementIntegralVariablePostprocessor) for areas where the variable value is not equal to 1. Since I need the area of my marker area, so I want to know this.
```
[AuxVariables]
  [fusion_region]
    order = FIRST
    family = LAGRANGE
  []
[]

[AuxKernels]
  [fusion_indicatorAuxKernel]
    type = ParsedAux
    variable = fusion_region
    coupled_variables = 'gr0 gr1 bnds c'
    expression = 'if(gr0 > 0.45, if(gr0 < 0.55, if(gr1 > 0.45, if(gr1 < 0.55, 1, 0), 0), 0), 0)'
    execute_on = 'initial timestep_end'
  []
[]

[Postprocessors]
  [fusion_area]
    type = ElementIntegralVariablePostprocessor
    variable = fusion_region
    execute_on = 'initial timestep_end'
  []
[]
```
Perhaps, for me, I want to mark an area in this way, and then calculate its area, which is better? Look forward to your reply and help, thank you.

**Reply by GiudGiud** [reply]
if you want the area of every element which has a nonzero value, you should create a:
- a new CONSTANT monomial variable to store the criterion
- use a ParsedAux with logic based on the current `fusion_region` values. something like this for the expression
```
(fusion_region > 0, 1, 0)
```

**Reply by HPL001** [reply]
Thank you for your reply. I have made the following Settings. My target is to get the area of the free surface of two particles. A free surface is defined as c between 0.45 and 0.55. 
```
[AuxVariables]
  [free_surface_region]
    order = CONSTANT
    family = MONOMIAL
  []

  [free_surface_int]
    order = CONSTANT
    family = MONOMIAL
  []
[]

[AuxKernels]
  [free_surface_indicatorAuxKernel]
    type = ParsedAux
    variable = free_surface_region
    coupled_variables = 'gr0 gr1 c'
    expression = 'if(c > 0.45, if(c < 0.55, 1, 0), 0)'
    execute_on = 'initial timestep_end'
  []

  [free_surface_indicator_intAuxKernel]
    type = ParsedAux
    variable = free_surface_int
    coupled_variables = 'free_surface_region'
    expression = 'if(free_surface_region > 0, 1, 0)'
    execute_on = 'initial timestep_end'
  []
[]

[ICs]
  [ic_gr0]
    type = SmoothCircleIC
    invalue = 1.0
    outvalue = 0.0
    radius = 7.5
    variable = gr0
    x1 = 12.5
    y1 = 10
    int_width = 1.9
  []

  [ic_gr1]
    type = SmoothCircleIC
    invalue = 1.0
    outvalue = 0.0
    radius = 7.5
    variable = gr1
    x1 = 27.5
    y1 = 10
    int_width = 1.9
  []

  [ic_c]
    type = SpecifiedSmoothCircleIC
    invalue = 1.0
    outvalue = 0.0
    radii = '7.5 7.5'
    variable = c
    x_positions = '12.5 27.5'
    y_positions = '10.0 10.0'
    z_positions = '0.0 0.0'
    int_width = 1.9
    3D_spheres = false
    block = 0
  []
[]

[Postprocessors]
  [free_surface_area_int]
    type = ElementIntegralVariablePostprocessor
    variable = free_surface_int
    execute_on = 'initial timestep_end'
  []

  [free_surface_int_area_initial]
    type = ElementIntegralVariablePostprocessor
    variable = free_surface_int
    execute_on = 'initial'
    # outputs = none
  []

  [free_surface_int_area_relative_percentage]
    type = ParsedPostprocessor
    expression = '100 * (free_surface_area_int - free_surface_int_area_initial) / free_surface_int_area_initial'
    pp_names = 'free_surface_area_int free_surface_int_area_initial'
  []
[]

[Executioner]
  type = Transient
  solve_type = 'NEWTON'
  l_tol = 1e-05
  l_max_its = 30
  nl_max_its = 15
  nl_rel_tol = 1e-04
  nl_abs_tol = 1e-04
  start_time = 0.0
  end_time = 100.0
  dt = 0.1
  [Adaptivity]
    initial_adaptivity = 2
    refine_fraction = 0.8 
    coarsen_fraction = 0.05 
    max_h_level = 3
  []
[]

[Outputs]
  exodus = true
  print_linear_residuals = true
  execute_on = "initial timestep_end"
  checkpoint = true
  solution_history = true
[]
```
I'm not sure if I'm setting this up correctly, but I'm having a few problems. For example, my free surface varies significantly more than reported in the literature. In addition, according to my initialization conditions, the initial free surface should be the circumference of two spherical particles, but I encountered the following situation.
![image](https://github.com/user-attachments/assets/a96a6ca4-10f1-4a57-8a0f-46871004acc4)
![image](https://github.com/user-attachments/assets/3342a5e2-dee0-41ef-aedb-df4766cc1555)
It looks like there are some mistakes. Can you give me some advice?

**Reply by GiudGiud** [reply]
This isn't the right approach to measure a perimeter or a surface. To do that you would need to run a line through the elements that mark the boundary, and measure its length. 

I think there is already machinery to measure surfaces of grains in the phase field module. Did you try to find that?

**Reply by HPL001** [reply]
I looked up the Complete Syntax, but I didn't find any related app for free surfaces, could you provide some references

**Reply by GiudGiud** [reply]
Ok you'll need to make an object then. 

I would say to try a user object that evaluated the true path of that line (in 2D) / surface (in 3D) by first placing its points in the middle of the elements with a nonzero value (or at the maximum?)

**Reply by HPL001** [reply]
I found an object, GrainBoundaryArea. Does it help me with my needs? I tried to experiment with him, and my two particles had an initial diameter of 15, but he output 47.1, which is the circumference of a particle.
```
  [bnd_length] 
    type = GrainBoundaryArea
    grains_per_side = 2.0
    execute_on = 'initial timestep_end'
  []
```

**Reply by GiudGiud** [reply]
47 could be 15 * pi
And the factor of 2 is in the grains per side

I think this is the right object

**Reply by HPL001** [reply]
Thanks for your reply, so does grains_per_side = 2.0 mean total grain boundary length divided by 2? In addition, when two particles touch, can the contact area (circled in the graph) be included by GrainBoundaryArea?
![image](https://github.com/user-attachments/assets/16bae39c-8b25-4488-91f3-b7cc307f5366)

**Reply by GiudGiud** [reply]
I just checked the definition and no it's not that.
You ll want to experiment with it to compute the contact area. You might need to modify it
